### DISCUSSION TITLE
Remap all Alt-* keymaps

### DISCUSSION THREAD

**Discussion by nik-rev** [helix-editor/helix#12030] [Ideas]
Alt keymappings are problematic because many people don't expect programs to create them. Often times alt keys are used for window managers or are specifically controlled by the user

This discussion happened in the past and people seem to not prefer it either: https://github.com/helix-editor/helix/discussions/6853

This discussion can serve as a place to brainstorm where the Alt-* keymappings should be remapped. Below is a table of all the alt keymappings.

| Keybinding   | Action                              |
|--------------|-------------------------------------|
| A-.          | repeat_last_motion                 |
| A-`          | switch_to_uppercase                |
| A-d          | delete_selection_noyank            |
| A-c          | change_selection_noyank            |
| A-C          | copy_selection_on_prev_line        |
| A-s          | split_selection_on_newline         |
| A-minus      | merge_selections                   |
| A-_          | merge_consecutive_selections       |
| A-;          | flip_selections                    |
| A-o / A-up   | expand_selection                   |
| A-i / A-down | shrink_selection                   |
| A-I          | select_all_children                |
| A-S-down     | select_all_children                |
| A-p / A-left | select_prev_sibling                |
| A-n / A-right| select_next_sibling                |
| A-e          | move_parent_node_end               |
| A-b          | move_parent_node_start             |
| A-a          | select_all_siblings               |
| A-x          | shrink_to_line_bounds              |
| A-J          | join_selections_space              |
| A-K          | remove_selections                  |
| A-,          | remove_primary_selection           |
| A-(          | rotate_selection_contents_backward |
| A-)          | rotate_selection_contents_forward  |
| A-:          | ensure_selections_forward          |
| A-C          | toggle_block_comments              |
| A-!          | shell_append_output                |
| `A-|`          | `shell_pipe_to`                      |
| A-u          | earlier                            |
| A-U          | later                              |

I'm currently just remapping all of them under `M`, because `M` at the moment has no functionality. What are your opinions?

**Comment by sicher** [comment]
I think it would be worthwhile to look into eliminating alt-keys. Moving these into one of more minor modes would help discoverability greatly.

**Comment by nik-rev** [comment]
> I think it would be worthwhile to look into eliminating alt-keys. Moving these into one of more minor modes would help discoverability greatly.

Almost all Alt keymappings manipulate selections in one way or another

Maybe a good way to get started would be to take all these selection-manipulation api and put them under "Tab" key? That key doesnt seem to do anything in normal mode so it could potentially be repurposed

the reason I'm suggest Tab and not something like M, is that tab is just 1 keystroke as opposed to M which is 2 keystrokes and a bit of stretching

Thoughts?

**Comment by sicher** [comment]
The 'm' key currently contains a few "match" oriented commands, but not that many. Another option is to rename it "modify" or "manipulate" and populate it with these. The one current thing under 'm' that would not make much sense is "goto matching bracket".

**Reply by nik-rev** [reply]
> The 'm' key currently contains a few "match" oriented commands, but not that many. Another option is to rename it "modify" or "manipulate" and populate it with these. The one current thing under 'm' that would not make much sense is "goto matching bracket".

let me know what you think about remapping Alt-C to just D: https://github.com/helix-editor/helix/pull/12072

I feel like it makes sense to make these PRs one at a time to gradually phase out Alt keys, it's hard to all agree on what should be so I'm making a pull request as a place where we can discuss that for one specific keymapping

**Comment by kumawatdarshan** [comment]
I dont specifically have conflicting binds but i find these quite hard to remember, for how useful some are.

Like you said about using tab as minor mode seems a nice idea but some of these keybinds are TS specific while others are more generic. Would definitely prefer grouping all the TS binds in its own minor mode as it also brings up the which key menu. 

For the other, more generic ones, I particularly dont have anything to add.

**Comment by sdoerner** [comment]
I've been running with the [super branch](https://github.com/helix-editor/helix/tree/super) and been using Cmd+key (on Mac) instead of Alt+Key for many of these. I really like this option. I also used that technique for various Kakoune-style bindings.

Terminal support is not as good as I hoped so far, but it works in Alacritty (presumably also Kitty). Wezterm is a bit [buggy](https://github.com/wez/wezterm/issues/5193#issuecomment-2387090568).

**Comment by dedebenui** [comment]
Mac OS uses the alt key for special characters inputs in some languages. For example, on my keyboard, `alt+e` yields `€`, `alt+n` yields `~`, etc. I need to be able to type these characters so I cannot use any alt keybindings.

**Comment by ysooqe** [comment]
Alt-* remappings to get rid of any Alt keybindings would be amazing! Putting specific "groups" of current Alt-bindings, like selection manipulation, in a separate "minor mode" sounds like a reasonable approach.

@nik-rev , the initial table seems to be missing some currently non-remappable Alt keybindings like the ones to be used [within a prompt](https://docs.helix-editor.com/keymap.html#prompt) (`Alt-Backspace`, `Alt-Delete`, etc.) or [in a picker](https://docs.helix-editor.com/keymap.html#picker) (`Alt-Enter`). Ideally, these would be remapped aswell?

**Comment by krzysztofkusmierczyk** [comment]
Alt is used to type acctented characters in some languages. Binding all Alt-* commands to something else is must have for me.

**Comment by tomekp1981** [comment]
I'm all for remapping to a minor mode. In my case, alt key is taken over by 3d applications (alt-key viewport manipulation, often non-remappable), so desktop/window manager/global operations (like switching between desktops, moving or scaling windows) use meta key. In this situation, minor mode is the best solution imo. Unless we get keyboards with another modifier key - that would be awesome :)
